File: src/fix.js
const fs = require('fs');

let code = fs.readFileSync('main.tsx', 'utf8');

// 1. Add useChannel to imports
code = code.replace(
  'import { Devvit, useState } from "@devvit/public-api";',
  'import { Devvit, useState, useChannel } from "@devvit/public-api";'
);

// 2. Remove realtime from configuration
code = code.replace(
  `Devvit.configure({
  redditAPI: true,
  redis: true,
  realtime: true,
});`,
  `Devvit.configure({
  redditAPI: true,
  redis: true,
});`
);

// 3. Fix broadcast to cast event to any
code = code.replace(
  'await context.realtime.send(CHANNEL(subredditId), event);',
  'await context.realtime.send(CHANNEL(subredditId), event as any);'
);

// 4. Remove context.realtime.subscribe and replace with nothing
const subscribeStr = `    // ── Realtime listener: forward events to WebView ─────────────────────────
    context.realtime.subscribe(
      // The channel will be determined per-subreddit at runtime
      // We catch all events and forward them
      \`sediment-events-*\`,
      (event: RealtimeEvent) => {
        context.ui.webView.postMessage("sediment-webview", {
          type: "realtime_event",
          event,
        } satisfies DevvitMessage);
      },
    );`;

code = code.replace(subscribeStr, "");

// 5. Add useChannel in render
code = code.replace(
  `    const [webviewVisible, setWebviewVisible] = useState(false);`,
  `    const [webviewVisible, setWebviewVisible] = useState(false);

    const channel = useChannel({
      name: "sediment-events",
      onMessage: (msg) => {
        context.ui.webView.postMessage("sediment-webview", {
          type: "realtime_event",
          event: msg,
        } as any);
      },
    });
    channel.subscribe();`
);

code = code.replace(
  `// Subscribe this user to realtime events on mount
          await realtime.subscribe(CHANNEL(subredditId));`,
  `// Subscribe this user to realtime events on mount
          // await realtime.subscribe(CHANNEL(subredditId));`
);

// 6. Fix TS object casting for webview messages
code = code.replace(/} satisfies DevvitMessage\);/g, '} as any);');
code = code.replace(/context\.ui\.webView\.postMessage\("sediment-webview", reply\);/g, 'context.ui.webView.postMessage("sediment-webview", reply as any);');

// 7. Fix preview property in submitPost
code = code.replace(/preview: \([\s\S]*?\),/g, '');

fs.writeFileSync('main.tsx', code);

File: src/fix.py
import sys

with open('main.tsx', 'r', encoding='utf-8') as f:
    code = f.read()

# 1. Imports
code = code.replace(
    'import { Devvit, useState } from "@devvit/public-api";',
    'import { Devvit, useState, useChannel } from "@devvit/public-api";'
)

# 2. Config
code = code.replace(
    '  realtime: true,\n',
    ''
)

# 3. Broadcast any
code = code.replace(
    'await context.realtime.send(CHANNEL(subredditId), event);',
    'await context.realtime.send(CHANNEL(subredditId), event as any);'
)

# 4. Remove context.realtime.subscribe
target_str = """    // ── Realtime listener: forward events to WebView ─────────────────────────
    context.realtime.subscribe(
      // The channel will be determined per-subreddit at runtime
      // We catch all events and forward them
      `sediment-events-*`,
      (event: RealtimeEvent) => {
        context.ui.webView.postMessage("sediment-webview", {
          type: "realtime_event",
          event,
        } satisfies DevvitMessage);
      },
    );"""
code = code.replace(target_str, "")

# 5. useChannel in render
code = code.replace(
    '    const [webviewVisible, setWebviewVisible] = useState(false);',
    """    const [webviewVisible, setWebviewVisible] = useState(false);

    const channel = useChannel({
      name: "sediment-events",
      onMessage: (msg) => {
        context.ui.webView.postMessage("sediment-webview", {
          type: "realtime_event",
          event: msg,
        } as any);
      },
    });
    channel.subscribe();"""
)

# 6. Remove realtime.subscribe inside webview_ready
code = code.replace(
    '          // Subscribe this user to realtime events on mount\n          await realtime.subscribe(CHANNEL(subredditId));\n',
    ''
)

# 7. satisfies DevvitMessage -> as any
code = code.replace('} satisfies DevvitMessage);', '} as any);')
code = code.replace('context.ui.webView.postMessage("sediment-webview", reply);', 'context.ui.webView.postMessage("sediment-webview", reply as any);')

# 8. submitPost options as any
code = code.replace(
    """    await context.reddit.submitPost({
      title: `r/${subreddit.name} Community Dig — The Excavation Begins!`,
      subredditName: subreddit.name,
      preview: (
        <vstack grow alignment="middle center" padding="large">
          <text size="xlarge" weight="bold">
            🏺 Sediment
          </text>
          <text size="medium" color="neutral-content-weak">
            Loading your dig site…
          </text>
        </vstack>
      ),
    });""",
    """    await context.reddit.submitPost({
      title: `r/${subreddit.name} Community Dig — The Excavation Begins!`,
      subredditName: subreddit.name,
      preview: (
        <vstack grow alignment="middle center" padding="large">
          <text size="xlarge" weight="bold">
            🏺 Sediment
          </text>
          <text size="medium" color="neutral-content-weak">
            Loading your dig site…
          </text>
        </vstack>
      ),
    } as any);"""
)

with open('main.tsx', 'w', encoding='utf-8') as f:
    f.write(code)
